appmap/_implementation/recorder.py

```diff
@@ -1,81 +1,184 @@
         logger.debug("AppMap recording started")
-        if self.enabled:
+        if self._enabled:
             logger.error("Recording already in progress, previous start:")
             logger.error("".join(traceback.format_list(self.start_tb)))
             raise RuntimeError("Recording already in progress")
         self.start_tb = traceback.extract_stack()
-        self.enabled = True
+        self._enabled = True
 
-    def stop_recording(self):
+    @abstractmethod
+    def _stop_recording(self):
         logger.debug("AppMap recording stopped")
-        self.enabled = False
+        self._enabled = False
         self.start_tb = None
         return self._events
 
-    @classmethod
-    def add_event(cls, event: Event):
-        """
-        Add the given event to the global recorder, as well as the recorder for the thread
-        identified in the event.
+    @abstractmethod
+    def _add_event(self, event):
+        self._events.append(event)
+
+    @staticmethod
+    def _initialize():
+        """Create a new default, shared recorder.
+
+        This method is intentionally not thread-safe. It really doesn't make sense have multiple
+        threads initializing the default recorder. If you find yourself wanting to do that, you
+        should probably be using per-thread recording.
         """
-        for tid in [None, event.thread_id]:
-            if tid in cls._recorders:
-                cls._recorders[tid]._add_event(event)
+        global _default_recorder
+        _default_recorder = SharedRecorder()
 
-    def _add_event(self, event: Event):
-        self._events.append(event)
+
+class ThreadRecorder(Recorder):
+    """
+    A Recorder to use for a thread. Not thread-safe, of course.
+    """
 
     @property
     def events(self):
-        """
-        Get the events from the current recording
-        """
-        return self._events
+        return super().events
+
+    @property
+    def next_event_id(self):
+        return super().next_event_id
+
+    def _start_recording(self):
+        super()._start_recording()
+
+    def _stop_recording(self):
+        return super()._stop_recording()
+
+    def _add_event(self, event):
+        super()._add_event(event)
+
+
+class SharedRecorder(Recorder):
+    """
+    A shared Recorder. The global recorder is an instance of this class.
+    """
+
+    _lock = threading.Lock()
+
+    @property
+    def events(self):
+        with self._lock:
+            return super().events
+
+    def _start_recording(self):
+        with self._lock:
+            super()._start_recording()
+
+    def _stop_recording(self):
+        with self._lock:
+            return super()._stop_recording()
+
+    @property
+    def next_event_id(self):
+        with self._lock:
+            return super().next_event_id
+
+    def _add_event(self, event):
+        with self._lock:
+            super()._add_event(event)
 
 
 def initialize():
-    Recorder.initialize()
+    Recorder._initialize()
```

Comment:
It seems to me it doesn't do any initialization anymore that would require other things set up before. Should it just be created at top-level statically? Eg.
```python
_default_recorder = SharedRecorder()
```

This way you could get rid of all the ugly `initialize()` methods.

Reply:
The main reason for `initialize` is to reset the state of all the modules before each test case runs. I think that's still necessary, and I don't really want to take the time to investigate it for this PR.